[Java FFI][Bug 18845] LibFoundation and LibScript support for Java FFI

livecodeali opened the pull request:
This patch adds support for interfacing with the Java Native Interface via LCB. 

libfoundation: Virtual machine initialisation
An API has been added to libfoundation for loading dynamic libraries. This is used along with a tweak to the weak stubs generator script to enable weak linking to libraries in libfoundation in exactly the same way as it is done in the engine. A Java Virtual Machine stubs file has been added which exposes the functions needed to initialise a JVM, and this is used on Mac and Linux when requested.

libfoundation: Building
A few changes have been made to the gyp/environment configuration to support building libfoundation with Java support. `config.sh` now looks for the required java headers on all platforms in the usual places, and the relevant folders are added to the include search paths when compiling on Mac or Linux.

libfoundation: Java Support API
A few functions have been added to the libfoundation API to support manipulation of Java objects (eg conversion to and from MCStringRefs) and constructing Java objects. An `MCJavaObjectRef` opaque type has been added which wraps a jobject.

libscript: Java FFI
Support for Java foreign bindings has been added via the usual LCB foreign binding string. Java bindings are specified using `"java:className>methodName methodSignature"`, for example: 
```
java.lang.Object>getClass()Ljava/lang/Class;"
```

libscript: Java module
An LCB java support module has been added to libscript which wraps the libfoundation Java support API. It tries to initialise the JVM when any of its functionality is used.

Finally, a basic LCB tests have been added for round-tripping between strings and java strings, and creating an instance of the `Object` class.

At runtime, the jvm dynamic library is found using the `$JAVA_HOME` environment variable on Mac. On Linux, it is expected to be found using the `$LD_LIBRARY_PATH` environment variable.

livecodeali committed:
[[ Foundation ]] Add MCStringSubstringIsEqualToCString

livecodeali committed:
[[ Java FFI ]] Add foundation-java and initialise JVM on Mac

livecodeali committed:
[[ Java FFI ]] Add utility functions for Java FFI processing and execution

livecodeali committed:
[[ Java FFI ]] Add support to libscript for parsing java foreign handler binding

livecodeali committed:
[[ Java FFI ]] Use JNI to resolve java method ID and execute

livecodeali committed:
[[ Java FFI ]] Add java LCB utility functions to libscript

montegoulding commented on libfoundation/src/foundation-java.cpp:
```diff
@@ -0,0 +1,234 @@
+    vm_args.options = options;
+    vm_args.ignoreUnrecognized = false;
+    
+    JNI_CreateJavaVM(&s_jvm, (void **)&s_env, &vm_args);
```
this should probably be:

```
return JNI_CreateJavaVM(&s_jvm, (void **)&s_env, &vm_args) == JNI_OK;
```

livecodeali replied:
I wonder if the best thing to do would be to allow it to fail, and just keep a variable around like `s_java_initialised` so that everything can just return false if it failed

montegoulding commented on libfoundation/src/foundation-java.cpp:
```diff
@@ -0,0 +1,234 @@
+void __MCJavaFinalize()
+{
+#ifdef TARGET_PLATFORM_MACOS_X
+    s_jvm->DestroyJavaVM();
```
probably want to handle `JVM_CreateJavaVM` failure in all these also

montegoulding reviewed (commented):
This is cool! Will we be adding Linux and Windows support at this stage or is that not in the MVP?

livecodeali commented:
Just Android for the MVP, but adding Mac was good for testing locally and adding Linux will be good for testing remotely I suppose - not sure how difficult it would be to add support on Windows, maybe not very?

runrevmark commented on libfoundation/src/foundation-string.cpp:
```diff
@@ -2957,16 +2965,26 @@ bool MCStringSubstringIsEqualToSubstring(MCStringRef self, MCRange p_sub, MCStri
+MC_DLLEXPORT_DEF
+bool MCStringSubstringIsEqualToNativeChars(MCStringRef self, MCRange p_range, const char_t *p_chars, uindex_t p_char_count, MCStringOptions p_options)
+{
+    __MCAssertIsString(self);
```
I think you need to clamp the range somewhere before the native calls are invoked.

livecodeali replied:
Oh yes, thanks!

livecodeali committed:
[[ Java FFI ]] Get method params using handler signature

livecodeali committed:
[[ Java FFI ]] Fail to initialise if JVM creation errors

livecodeali committed:
[[ Java FFI ]] Get JVM running on Travis

livecodeali committed:
[[ Java FFI ]] Dynamically load jvm lib on mac

livecodeali committed:
[[ Java FFI ]] Initialise JVM on Linux
This patch attempts to initialise the JVM on startup on Linux by
loading libjvm.so from the expected location relative to the
JAVA_HOME environment variable

livecodeali committed:
[[ Java FFI ]] Allow JVM initialisation to fail
Use s_java_initialised variable to prevent attempted access to the
jni if jvm initialisation fails

livecodeali committed:
Merge branch 'develop' of github.com:livecode/livecode into feature-java_ffi_support

livecodeali committed:
[[ Java FFI ]] Add tests of basic functionality

livecodeali committed:
Merge branch 'develop' of github.com:livecode/livecode into feature-java_ffi_support

livecodeali committed:
Merge branch 'develop' of github.com:livecode/livecode into feature-java_ffi_support

livecodeali committed:
[[ New Operator ]] Replace new with new (nothrow)

livecodeali committed:
[[ LCB Java ]] Only initialise jvm on request in java module

livecodeali committed:
[[ Dynamic Java ]] Add load module api to libfoundation

livecodeali committed:
[[ Dynamic Java ]] Add parameter to stubs generator
Using --foundation allows the stubs generator to use the libfoundation
version of the module load/unload/resolve API

livecodeali committed:
[[ Dynamic Java ]] Use stubs-generated functions for jvm loading

livecodeali committed:
[[ Dynamic Java ]] Ensure path to libjvm is correct on Linux
Specifically, ensure it takes into account the architecture.

livecodeali committed:
[[ Dynamic Java ]] Add load module api to libfoundation

livecodeali committed:
[[ Dynamic Java ]] Add generated source code to libfoundation

livecodeali committed:
[[ Dynamic Java ]] Set JAVA_HOME env var on mac Travis build

livecodefraser commented on config/linux.gypi:
```diff
@@ -7,6 +7,7 @@
 		'objcopy': '<!(echo ${OBJCOPY:-objcopy})',
 		'objdump': '<!(echo ${OBJDUMP:-objdump})',
 		'strip':   '<!(echo ${STRIP:-strip})',
+		'javahome': '<!(echo ${JAVA_HOME:-javahome})',
```
The bash syntax `${FOO:-bar}` means "if FOO is set, use the value of it, otherwise default to bar"; I'm not sure defaulting JAVA_HOME here is particularly useful.

If you have a look at config.sh, it has some "well known" locations for Android on Linux - they could perhaps be moved out of the Android-specific section of that file and used to set JAVA_HOME if not already set on Linux.

livecodefraser commented on libfoundation/libfoundation.gyp:
```diff
@@ -106,10 +108,39 @@
+					],
+					'outputs':
+					[
+						'<(INTERMEDIATE_DIR)/src/libfoundationjvm.<(OS).stubs.cpp',
```
For future reference, we don't need to add `<(OS)` to the intermediate file name here - exiting bits that do so are just left-overs from the pre-Gyp days.

livecodefraser commented on libfoundation/src/foundation-java.cpp:
```diff
@@ -0,0 +1,932 @@
+#if defined(__X86_64__)
+	t_target = "/jre/lib/amd64/server/libjvm.so";
+#elif defined(__I386__)
+	t_target = "/jre/lib/i386/server/libjvm.so";
```
I don't suppose there's any alternative to hard-coding these JVM library paths?

livecodeali replied:
Not that I know of, other than using `system` :-(

peter-b replied:
Unfortunately this won't even work on most Linux installations, including my Fedora VM. 😞  `getenv()` for `JAVA_HOME` any good?

livecodeali replied:
That path is on top of a `getenv(JAVA_HOME)`

Why specifically won't it work?

livecodefraser commented on libfoundation/src/system-dylib-mac.cpp:
```diff
@@ -0,0 +1,73 @@
+	if (t_result != nil)
+		return t_result;
+	
+	CFURLRef t_url;
```
I think this repeats the same bug as the existing MacOS dylib code - it overloads the object handle as being both a `NSImage*` and `CFBundleRef` which will lead to crashes when passed to the wrong function. Do we actually need both the NS code and the CF code?

livecodeali replied:
I'll try removing the NS code and see what happens

livecodefraser commented on libfoundation/src/system-file-posix.cpp:
```diff
@@ -24,6 +24,11 @@ along with LiveCode.  If not see <http://www.gnu.org/licenses/>.  */
+#ifdef _MACOSX
+#include <mach-o/dyld.h>
+#endif
 
```
Why the requirement to include the dynamic loading stuff here?

livecodeali replied:
Just left over from when I first implemented the module loading stuff in that file :-) will remove

livecodefraser commented on util/weak_stub_maker.pl:
```diff
@@ -2,10 +2,27 @@
+    $resolve = "MCU_resolvemodulesymbol";
+}
+
 # Read all of the input data
```
Could we just change the weak stub maker to always use the libfoundation implementations instead of the MCU_* ones?

livecodeali replied:
Probably, yes!

livecodeali committed:
[[ Foundation ]] Call foundation dylib implementations from engine

livecodeali committed:
[[ Foundation ]] Add comment explaining purpose of non-engine-dependent weak stub maker

livecodeali committed:
[[ Foundation ]] Remove errant inclusion

livecodeali committed:
[[ Dynamic Java ]] Remove unnecessary os param from generated filename

livecodeali committed:
[[ Dynamic Java ]] Set javahome gyp variable more reliably

livecodeali committed:
[[ Foundation Dylib ]] Use NSImageAdd to load modules in libfoundation
This is used by the engine to try to load libraries on MacOS, and if
it failes, CFBundleCreate is used instead. This restores existing
behavior.

Note that this renders some additional MacOS fettling redundant, as
an attempt is already made by the engine to load modules relative
to the current executable path.

livecodeali committed:
[[ Dynamic Java ]] Look in LD_LIRBARY_PATH for libjvm.so

livecodeali commented:
Ok I think that's all the points addressed that it was possible to address. The NSImage / CFBundle thing is a pain, but I think it's no worse in the engine than it was before.

Here's a PR testing some of the inclusions which hopefully proves I haven't messed up dylib loading: https://github.com/livecode/livecode/pull/4934

livecodeali committed:
[[ Gyp ]] Don't error if JAVA_SDK is not found on windows

livecodeali committed:
Merge branch 'develop' of github.com:livecode/livecode into feature-java_ffi_support

Conflicts:
	config.sh

livecodefraser commented:
@livecode-vulcan review okay 04080d5

livecode-vulcan commented:
💙 review by @livecodefraser ok 04080d590854523542f284f1919dff32b2bddb55

livecode-vulcan commented:
😞 test failure 04080d590854523542f284f1919dff32b2bddb55

* try-community-linux-x86: [pending](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/786)
* try-community-mac: [failure](https://vulcan.livecode.com/builders/try-community-mac/builds/791)
* try-community-linux-x86_64: [pending](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/784)
* try-community-emscripten: [pending](https://vulcan.livecode.com/builders/try-community-emscripten/builds/785)
* try-community-ios: [failure](https://vulcan.livecode.com/builders/try-community-ios/builds/791)
* try-community-win-x86: [pending](https://vulcan.livecode.com/builders/try-community-win-x86/builds/783)
* try-community-android-armv6: [pending](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/783)

livecodeali commented:
Hmm, we may need to install a JDK on the build macs for this to work.

livecodeali commented:
@livecode-vulcan retry 04080d5

livecode-vulcan commented:
😞 test failure 04080d590854523542f284f1919dff32b2bddb55

* try-community-linux-x86: [pending](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/843)
* try-community-mac: [pending](https://vulcan.livecode.com/builders/try-community-mac/builds/849)
* try-community-linux-x86_64: [pending](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/841)
* try-community-emscripten: [failure](https://vulcan.livecode.com/builders/try-community-emscripten/builds/842)
* try-community-ios: [failure](https://vulcan.livecode.com/builders/try-community-ios/builds/849)
* try-community-win-x86: [pending](https://vulcan.livecode.com/builders/try-community-win-x86/builds/840)
* try-community-android-armv6: [pending](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/840)

livecodeali committed:
[[ Java FFI ]] Tweak preprocessor directives for cross-platform compile

livecodeali commented:
@livecode-vulcan try b6c1c02

livecodefraser commented:
@livecodevulcan review okay b6c1c02

livecodefraser commented:
@livecode-vulcan review okay b6c1c02
even ...

livecode-vulcan commented:
💙 review by @livecodefraser ok b6c1c0255825629e7d9b080377df59dce56a2af5

livecode-vulcan commented:
😞 test failure b6c1c0255825629e7d9b080377df59dce56a2af5

* try-community-linux-x86: [pending](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/861)
* try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/867)
* try-community-linux-x86_64: [pending](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/859)
* try-community-emscripten: [failure](https://vulcan.livecode.com/builders/try-community-emscripten/builds/860)
* try-community-ios: [failure](https://vulcan.livecode.com/builders/try-community-ios/builds/867)
* try-community-win-x86: [pending](https://vulcan.livecode.com/builders/try-community-win-x86/builds/858)
* try-community-android-armv6: [failure](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/858)

livecodeali committed:
[[ Java FFI Support ]] Tweak gyp files to exclude source files appropriately

livecodefraser commented:
@livecode-vulcan review okay 8b3271f

livecode-vulcan commented:
💙 review by @livecodefraser ok 8b3271ffdfb55a2cb16bf07fc309d56de8f24ff4

livecode-vulcan commented:
😞 test failure 8b3271ffdfb55a2cb16bf07fc309d56de8f24ff4

* try-community-linux-x86: [pending](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/867)
* try-community-mac: [pending](https://vulcan.livecode.com/builders/try-community-mac/builds/873)
* try-community-linux-x86_64: [pending](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/865)
* try-community-emscripten: [pending](https://vulcan.livecode.com/builders/try-community-emscripten/builds/866)
* try-community-ios: [pending](https://vulcan.livecode.com/builders/try-community-ios/builds/873)
* try-community-win-x86: [pending](https://vulcan.livecode.com/builders/try-community-win-x86/builds/864)
* try-community-android-armv6: [failure](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/864)

livecodeali committed:
[[ Java FFI Support ]] Location of java headers is dependent on host os

livecodefraser commented:
@livecode-vulcan review okay 74429d2

livecode-vulcan commented:
💙 review by @livecodefraser ok 74429d244593da7900e1e297ee3f7d50f0c55086

livecode-vulcan commented:
😞 test failure 74429d244593da7900e1e297ee3f7d50f0c55086

* try-community-linux-x86: [pending](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/869)
* try-community-mac: [pending](https://vulcan.livecode.com/builders/try-community-mac/builds/875)
* try-community-linux-x86_64: [pending](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/867)
* try-community-emscripten: [failure](https://vulcan.livecode.com/builders/try-community-emscripten/builds/868)
* try-community-ios: [failure](https://vulcan.livecode.com/builders/try-community-ios/builds/875)
* try-community-win-x86: [pending](https://vulcan.livecode.com/builders/try-community-win-x86/builds/866)
* try-community-android-armv6: [failure](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/866)

livecodeali committed:
[[ Java FFI Support ]] Only use javahome variable when appropriate
The java headers are needed in the following situations:
- building for mac (on mac)
- building for linux (on linux)
- building for android (on mac or linux)

This patch ensures that the javahome gyp variable is set in all
of these situations and that the relevant headers are found.

livecodefraser commented:
@livecode-vulcan review okay 2e24987

livecode-vulcan commented:
💙 review by @livecodefraser ok 2e24987919df1d8daf4405f5250f23c8d08d8509

livecode-vulcan commented:
😞 test failure 2e24987919df1d8daf4405f5250f23c8d08d8509

* try-community-linux-x86: [failure](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/870)
* try-community-mac: [pending](https://vulcan.livecode.com/builders/try-community-mac/builds/876)
* try-community-linux-x86_64: [failure](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/868)
* try-community-emscripten: [failure](https://vulcan.livecode.com/builders/try-community-emscripten/builds/869)
* try-community-ios: [failure](https://vulcan.livecode.com/builders/try-community-ios/builds/876)
* try-community-win-x86: [pending](https://vulcan.livecode.com/builders/try-community-win-x86/builds/867)
* try-community-android-armv6: [failure](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/867)

livecodeali committed:
[[ Java FFI Support ]] Move various functions into foundation-java-private
The TARGET_SUPPORTS_JAVA define is now used only once to make it
easier to see what's going on.

livecodeali committed:
[[ Java FFI Support ]] Set value of TARGET_SUPPORTS_JAVA in gyp file
This makes it be defined at the same time as adding the include
files for the JVM, so it's more obivous in what circustances java
ffi is available.

Also exclude system-dylib-mac from iOS builds.

livecodeali commented:
@livecode-vulcan review ok cd5eba2

livecodeali commented:
While this PR is WIP, I'm going to self review just to be able to try it on the builders.
Ultimately I am going to close this PR and reopen it based on whatever changes get merged in the foundation module loading branch

livecodeali commented:
@livecode-vulcan review ok cd5eba2

livecode-vulcan commented:
💙 review by @livecodeali ok cd5eba261d378b45f22cf8c853d91b1c34cb6963

livecode-vulcan commented:
😞 test failure cd5eba261d378b45f22cf8c853d91b1c34cb6963

* try-community-linux-x86: [pending](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/896)
* try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/902)
* try-community-linux-x86_64: [pending](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/894)
* try-community-emscripten: [pending](https://vulcan.livecode.com/builders/try-community-emscripten/builds/895)
* try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/902)
* try-community-win-x86: [pending](https://vulcan.livecode.com/builders/try-community-win-x86/builds/893)
* try-community-android-armv6: [failure](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/893)

livecodeali committed:
Use jvm stubs whenever needed

livecodeali commented:
@livecode-vulcan review ok a1539fb

livecode-vulcan commented:
💙 review by @livecodeali ok a1539fbda543ec90ace718af9812cf8266dc01a5

livecode-vulcan commented:
😞 test failure a1539fbda543ec90ace718af9812cf8266dc01a5

* try-community-linux-x86: [failure](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/897)
* try-community-mac: [pending](https://vulcan.livecode.com/builders/try-community-mac/builds/903)
* try-community-linux-x86_64: [failure](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/895)
* try-community-emscripten: [pending](https://vulcan.livecode.com/builders/try-community-emscripten/builds/896)
* try-community-ios: [pending](https://vulcan.livecode.com/builders/try-community-ios/builds/903)
* try-community-win-x86: [pending](https://vulcan.livecode.com/builders/try-community-win-x86/builds/894)
* try-community-android-armv6: [pending](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/894)

livecodeali committed:
Merge branch 'develop' of github.com:livecode/livecode into feature-java_ffi_support

livecodeali committed:
[[ Java FFI ]] Add JAVA_HOME detection on Linux

livecodeali commented:
@livecode-vulcan review ok bfdbe14
Trying this again. It still needs rebasing on the new library loading stuff in libfoundation.

livecode-vulcan commented:
💙 review by @livecodeali ok bfdbe14fb9e4abe4b4d47583c8b3bbf5868a2498

livecodeali commented:
Todo:
- [ ] Rebase on https://github.com/livecode/livecode/pull/5161
- [ ] Special case byte[] to convert to and from MCDataRef where necessary

livecode-vulcan commented:
😞 test failure bfdbe14fb9e4abe4b4d47583c8b3bbf5868a2498

* try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/947)
* try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/953)
* try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/945)
* try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/946)
* try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/953)
* try-community-win-x86: [failure](https://vulcan.livecode.com/builders/try-community-win-x86/builds/949)
* try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/944)
